[GTK][WPE] Use a single Damage object for damage in global coordinate space

Scony opened the pull request:
#### 022e8d04ae5fd197bb48f0b3f4695391be4c0507
<pre>
[GTK][WPE] Use a single Damage object for damage in global coordinate space
<a href="https://bugs.webkit.org/show_bug.cgi?id=296615">https://bugs.webkit.org/show_bug.cgi?id=296615</a>

Reviewed by Carlos Garcia Campos.

This change makes layers utilize global coordinate space damage object
originating from the layer tree host instead of per-layer instances.
This way, fewer damage rectangles are stored for global coordinate
space in total thus making the storing and iterating more efficient.

Canonical link: <a href="https://commits.webkit.org/298073@main">https://commits.webkit.org/298073@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9fa48609ba630a1241f0f09483ee3fd796487a82

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/114004 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/33756 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/24214 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/120172 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/64781 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/115893 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/34384 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/42317 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/86646 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/41666 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/116952 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/27372 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/102390 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/67029 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/26557 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/20521 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/63879 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/96737 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/20638 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/123397 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/41037 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/30563 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/95475 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/41415 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/98602 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/95252 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/40393 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/18200 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/37155 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/18297 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/40912 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/46414 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/40540 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/43839 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/42294 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

carlosgcampos commented on Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedPlatformLayer.h:
```diff
@@ -246,6 +247,7 @@ class CoordinatedPlatformLayer : public ThreadSafeRefCounted<CoordinatedPlatform
 
 #if ENABLE(DAMAGE_TRACKING)
     bool m_damagePropagationEnabled { false };
+    Damage* m_damageInGlobalCoordinateSpace;
```
```suggestion
    Damage* m_damageInGlobalCoordinateSpace { nullptr };
```

carlosgcampos commented on Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/LayerTreeHost.cpp:
```diff
@@ -416,6 +420,8 @@ void LayerTreeHost::attachLayer(CoordinatedPlatformLayer& layer)
 #if ENABLE(DAMAGE_TRACKING)
     layer.setDamagePropagationEnabled(webPage().corePage()->settings().propagateDamagingInformation());
+    if (m_damageInGlobalCoordinateSpace)
+        layer.setDamageInGlobalCoordinateSpace(&*m_damageInGlobalCoordinateSpace);
```
This is problematic, the layer can be destroyed after the layer tree host. And we end up setting this pointer to the TextureMapperLayer in the compositing thread too.

Scony replied:
Changed optional/rawptr to `shared_ptr` - this way the managed object will be destroyed either with layer tree host, or last layer standing. As for the threading, the actual usage of managed object happens only in TextureMapperLayer and is coordinated so we should be good.

carlosgcampos commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
```diff
@@ -574,7 +564,7 @@ void TextureMapperLayer::damageWholeLayerIncludingItsRectFromPreviousFrame()
     // and re-appears in another. Therefore the damaging of a layer in the "old" state is required as well.
     if (m_previousLayerRectInGlobalCoordinateSpace)
-        ensureDamageInGlobalCoordinateSpace().add(*m_previousLayerRectInGlobalCoordinateSpace);
+        m_damageInGlobalCoordinateSpace->add(*m_previousLayerRectInGlobalCoordinateSpace);
```
Don't we need to nullcheck here?

Scony replied:
It's guaranteed it will be set if the propagation is enabled - I've added assert anyway to show the intent.

carlosgcampos commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
```diff
@@ -1274,7 +1264,7 @@ void TextureMapperLayer::addChild(TextureMapperLayer* childLayer)
 {
     if (node.m_previousLayerRectInGlobalCoordinateSpace)
-        ensureDamageInGlobalCoordinateSpace().add(*node.m_previousLayerRectInGlobalCoordinateSpace);
+        m_damageInGlobalCoordinateSpace->add(*node.m_previousLayerRectInGlobalCoordinateSpace);
```
And here

Scony replied:
Same

Scony committed:
[GTK][WPE] Use a single Damage object for damage in global coordinate space
https://bugs.webkit.org/show_bug.cgi?id=296615

Reviewed by Carlos Garcia Campos.

This change makes layers utilize global coordinate space damage object
originating from the layer tree host instead of per-layer instances.
This way, fewer damage rectangles are stored for global coordinate
space in total thus making the storing and iterating more efficient.

Canonical link: https://commits.webkit.org/298073@main
